search/controllers/api/classic_parser.py

```diff
@@ -0,0 +1,140 @@
+    open paren groups.
+2.  If a paren is closed, we pop out the most recent open paren from the stack
+    of start positions and recursively parse the paren group, adding it to our
+    list of tokens.
+3.  If a quote is opened, flip a quotation flag.
+4.  If a space is encountered, append the token to the list of tokens and reset
+    the token start position.
+
+Tokens are themselves parsed using two helpers: :func:`_parse_operator` and
+:func:`_parse_field_query`.
+
+The final, parsed query is a :class:`domain.api.Phrase`, which is a nested
+set of Tuples::
+
+    >>> parse_classic_query("au:del_maestro AND ti:checkerboard")
+    ((Field.Author, 'del_maestro'), Operator.AND, (Field.Title, 'checkerboard'))
+
+See :module:`tests.test_classic_parser` for more examples.
+"""
+
+
+from typing import Any, List, Optional, Tuple, Union
+
+from werkzeug.exceptions import BadRequest
+
+from ...domain.api import Phrase, Operator, Field
+
+
+def parse_classic_query(query: str) -> Phrase:
+    """
+    Parse Classic API-style query string into app-native Phrase.
+
+    Iterates through each character in the string, applying a recursive-descent
+    parser and appropriately handling parens and quotes. See module docstring.
+
+    Parameters
+    ----------
+    query : str
+        A Classic API query string.
+
+    Returns
+    -------
+    :class:`Phrase`
+        A tuple representing the query.
+    """
+
+    # Intializing state variables
+    tokens: List[Union[str, Phrase]] = []
+    token_start = 0
+    paren_group = []
+    in_quote = False
+
+    # Iterate through characters
+    for i, c in enumerate(query):
+        if c == '(':
+            paren_group.append(i) # add open paren start position to stack
+        elif c == ')':
+            start = paren_group.pop() # get innermost open paren
+            if not paren_group and start == 0 and i + 1 == len(query):
+                # Parent spans whole group, strip the parens and just return the de-parened phrase
+                return parse_classic_query(query[1:i])
+            elif not paren_group:
+                # pass the paren-stripped phrase for parsing
+                tokens.append(parse_classic_query(query[start+1:i]))
+                token_start = i+1
+        elif c == '"':
+            in_quote = not in_quote # flip quotation bit
+        elif c == ' ':
+            if in_quote:
+                continue # keep moving if parsing a quote
+            elif not paren_group and token_start != i:
+                tokens.append(query[token_start:i]) # append the token
+                token_start = i + 1
+            elif token_start == i:
+                token_start = i + 1 # multiple spaces, move the token_start back
+        else:
+            continue
+
+    # handle final-position token
+    if query[token_start:]:
+        tokens.append(query[token_start:])
+
+    # cast tokens to class-based representations
+    classed_tokens = []
```

Comment:
Consider something like this to make it quite clear what's going on (you'll probably need it when you decompose, anyway).

```suggestion
    classed_tokens: List[Union[Term, Operator, Phrase]] = []
```

Reply:
This has now been implemented!